[quant][qat] Ensure observer respects device affinity

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47514 [quant][qat] Ensure observer respects device affinity**

Summary:

fixed a bug where we were always setting the device to 'cuda' (irrespective of the device id)
in the calculate_qparams function

Fixes #46533

Test Plan:
python test/test_quantization.py TestObserver.test_observer_qparams_respects_device_affinity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24800495](https://our.internmc.facebook.com/intern/diff/D24800495)

supriyar committed:
[quant][qat] Ensure observer respects device affinity

Summary:
Previosuly the scale and zero_point were returned on the CPU even if
the input tensor was on the GPU.
This is because `copy_()` doesn't respect the device when copying over the tensor.

Also fixed a bug where we were always setting the device to 'cuda' (irrespective of the device id)
in the calculate_qparams function

Test Plan:
python test/test_quantization.py TestObserver.test_observer_qparams_respects_device_affinity

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
QAT Benchmark on MobilenetV2 
Before change: 
```
PyTorchObserver {"type": "_", "metric": "qnnpack_fp_latency_ms", "unit": "ms", "value": "229.25734519958496"}
PyTorchObserver {"type": "_", "metric": "qnnpack_qat0_latency_ms", "unit": "ms", "value": "522.6466655731201"}
PyTorchObserver {"type": "_", "metric": "qnnpack_qat1_obs_0_latency_ms", "unit": "ms", "value": "261.6147994995117"}
PyTorchObserver {"type": "_", "metric": "qnnpack_qat2_obs_0_bn_0_latency_ms", "unit": "ms", "value": "250.9593963623047"}
PyTorchObserver {"type": "_", "metric": "qnnpack_fp_cuda_max_mem_mib", "unit": "MiB", "value": "246.51806640625"}
PyTorchObserver {"type": "_", "metric": "qnnpack_qat0_cuda_max_mem_mib", "unit": "MiB", "value": "314.82080078125"}
PyTorchObserver {"type": "_", "metric": "qnnpack_qat1_obs_0_cuda_max_mem_mib", "unit": "MiB", "value": "314.82080078125"}
PyTorchObserver {"type": "_", "metric": "qnnpack_qat2_obs_0_bn_0_cuda_max_mem_mib", "unit": "MiB", "value": "314.66650390625"}
/home/supriyar/local/fbsource/fbcode/buck-out/dev/gen/caffe2/caffe2/fb/high_perf_models/pytorch/mobilenetv2/mobilenetv2_bench#link-tree/torch/quantization/observer.py:123: UserWarning: Please use quant_min and quant_max to specify the range for observers.                     reduce_range will be deprecated in a future release of PyTorch.
  reduce_range will be deprecated in a future release of PyTorch."
PyTorchObserver {"type": "_", "metric": "fbgemm_fp_latency_ms", "unit": "ms", "value": "231.16016387939453"}
PyTorchObserver {"type": "_", "metric": "fbgemm_qat0_latency_ms", "unit": "ms", "value": "546.6287136077881"}
PyTorchObserver {"type": "_", "metric": "fbgemm_qat1_obs_0_latency_ms", "unit": "ms", "value": "282.5195789337158"}
PyTorchObserver {"type": "_", "metric": "fbgemm_qat2_obs_0_bn_0_latency_ms", "unit": "ms", "value": "268.3393955230713"}
PyTorchObserver {"type": "_", "metric": "fbgemm_fp_cuda_max_mem_mib", "unit": "MiB", "value": "246.80126953125"}
PyTorchObserver {"type": "_", "metric": "fbgemm_qat0_cuda_max_mem_mib", "unit": "MiB", "value": "315.10400390625"}
PyTorchObserver {"type": "_", "metric": "fbgemm_qat1_obs_0_cuda_max_mem_mib", "unit": "MiB", "value": "315.10400390625"}
PyTorchObserver {"type": "_", "metric": "fbgemm_qat2_obs_0_bn_0_cuda_max_mem_mib", "unit": "MiB", "value": "314.94970703125"}
```

After change 
```
PyTorchObserver {"type": "_", "metric": "qnnpack_fp_latency_ms", "unit": "ms", "value": "221.08221054077148"}
PyTorchObserver {"type": "_", "metric": "qnnpack_qat0_latency_ms", "unit": "ms", "value": "506.84404373168945"}
PyTorchObserver {"type": "_", "metric": "qnnpack_qat1_obs_0_latency_ms", "unit": "ms", "value": "261.29603385925293"}
PyTorchObserver {"type": "_", "metric": "qnnpack_qat2_obs_0_bn_0_latency_ms", "unit": "ms", "value": "239.6218776702881"}
PyTorchObserver {"type": "_", "metric": "qnnpack_fp_cuda_max_mem_mib", "unit": "MiB", "value": "246.51806640625"}
PyTorchObserver {"type": "_", "metric": "qnnpack_qat0_cuda_max_mem_mib", "unit": "MiB", "value": "314.82080078125"}
PyTorchObserver {"type": "_", "metric": "qnnpack_qat1_obs_0_cuda_max_mem_mib", "unit": "MiB", "value": "314.82080078125"}
PyTorchObserver {"type": "_", "metric": "qnnpack_qat2_obs_0_bn_0_cuda_max_mem_mib", "unit": "MiB", "value": "314.66650390625"}
/home/supriyar/local/fbsource/fbcode/buck-out/dev/gen/caffe2/caffe2/fb/high_perf_models/pytorch/mobilenetv2/mobilenetv2_bench#link-tree/torch/quantization/observer.py:123: UserWarning: Please use quant_min and quant_max to specify the range for observers.                     reduce_range will be deprecated in a future release of PyTorch.
  reduce_range will be deprecated in a future release of PyTorch."
PyTorchObserver {"type": "_", "metric": "fbgemm_fp_latency_ms", "unit": "ms", "value": "214.65778350830078"}
PyTorchObserver {"type": "_", "metric": "fbgemm_qat0_latency_ms", "unit": "ms", "value": "502.2757053375244"}
PyTorchObserver {"type": "_", "metric": "fbgemm_qat1_obs_0_latency_ms", "unit": "ms", "value": "304.18944358825684"}
PyTorchObserver {"type": "_", "metric": "fbgemm_qat2_obs_0_bn_0_latency_ms", "unit": "ms", "value": "258.211612701416"}
PyTorchObserver {"type": "_", "metric": "fbgemm_fp_cuda_max_mem_mib", "unit": "MiB", "value": "246.80126953125"}
PyTorchObserver {"type": "_", "metric": "fbgemm_qat0_cuda_max_mem_mib", "unit": "MiB", "value": "315.10400390625"}
PyTorchObserver {"type": "_", "metric": "fbgemm_qat1_obs_0_cuda_max_mem_mib", "unit": "MiB", "value": "315.10400390625"}
PyTorchObserver {"type": "_", "metric": "fbgemm_qat2_obs_0_bn_0_cuda_max_mem_mib", "unit": "MiB", "value": "314.94970703125"}
```

vkuzo commented:
wow, interesting, great catch

> self.min_val = self.min_val.copy_(min_val).to(min_val.device)

is the `self.min_val.copy_(min_val).to(min_val.device)` code actually doing CUDA -> CPU -> CUDA?  If yes, is there a way to remove that?

supriyar commented:
> wow, interesting, great catch
> 
> > self.min_val = self.min_val.copy_(min_val).to(min_val.device)
> 
> is the `self.min_val.copy_(min_val).to(min_val.device)` code actually doing CUDA -> CPU -> CUDA? If yes, is there a way to remove that?

Hmm, it might. I think I can use `y = x.cuda(device=x.device)` instead to copy to same device if the input is a cuda tensor.

ngimel commented on torch/quantization/observer.py:
```diff
@@ -394,10 +394,12 @@ def forward(self, x_orig):
         max_val = torch.max(max_val_cur, self.max_val)
-        self.min_val.copy_(min_val)
-        self.max_val.copy_(max_val)
+        self.min_val = self.min_val.copy_(min_val)
```
You can do `self.min_val = min_val` I think? `self.min_val` already has the same device and dtype as minval, so there's no point in actually making copy.

vkuzo replied:
> You can do `self.min_val = min_val` I think? `self.min_val` already has the same device and dtype as minval, so there's no point in actually making copy.

unfortunately that doesn't work well with `nn.DataParallel`, as it assumes that a model's buffers do not change their memory addresses.  We rewrote it to explicitly copy into the same existing buffer in order to make `DataParallel` work properly.

vkuzo commented:
> I can instead to the to call first before the copy_ to make sure we move the tensor to CUDA before doing the COPY. @vkuzo thoughts on this approach?

that would make sense to me, and the performance metrics don't show a slowdown which is good.

Doesn't have to block this bugfix, but I think in the future it might make sense to rethink how we do device management in observers.  Ideally we can enforce strong assumptions that all observer buffers are on the correct device after initialization and any `to` calls, and then just throw an error with a stack trace when any mismatch is detected.  Then we could remove all the custom code which moves tensors between devices.

supriyar committed:
Update on "[quant][qat] Ensure observer respects device affinity"


Summary:
Previosuly the scale and zero_point were returned on the CPU even if
the input tensor was on the GPU.
This is because `copy_()` doesn't respect the device when copying over the tensor.

Also fixed a bug where we were always setting the device to 'cuda' (irrespective of the device id)
in the calculate_qparams function

Test Plan:
python test/test_quantization.py TestObserver.test_observer_qparams_respects_device_affinity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24800495](https://our.internmc.facebook.com/intern/diff/D24800495)

[ghstack-poisoned]

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -394,10 +394,12 @@ def forward(self, x_orig):
         x = x_orig.detach()  # avoid keeping autograd tape
         x = x.to(self.min_val.dtype)
         min_val_cur, max_val_cur = torch._aminmax(x)
+        self.min_val = self.min_val.to(x.device)
+        self.max_val = self.max_val.to(x.device)
```
does the `DataParallel` test past with this?  Would something like

```
self.min_val.copy_(min_val.to(x.device))
```

work instead?

supriyar replied:
Yes the tests in `TestDistributed` passed with this. 
MinMaxObserver is a little different and needs the `.to` here since we call `min_val = torch.min(min_val_cur, self.min_val)` after this which will fail since `self.min_val` and `min_val_cur` are on different devices.

vkuzo replied:
> self.min_val and min_val_cur are on different devices.

do we know why this happens, btw?  If the model owner does

```
inputs = inputs.cuda()
model = model.cuda()
model(inputs)
```

we would expect that all observes in model are already on the right device.

btw, it doesn't need to block this PR, I'm just wondering about the root cause for future improvements.

supriyar replied:
I believe it mainly happens when we use the observers standalone and they are not part of a model. Not sure if that is a use-case we want to support.
We can add a device argument to observers, which will ensure all the buffers registered in the observer are on the expected device (i.e. match the device of the input it is run on).

vkuzo replied:
> I believe it mainly happens when we use the observers standalone and they are not part of a model. Not sure if that is a use-case we want to support.

> Previosuly the scale and zero_point were returned on the CPU even if

Hmm, sorry, so is the goal of the PR to support this standalone use case?  I.e. the user created the observer on CPU, passes a CUDA Tensor, and this is fixing the crash?  If yes, it might be worth deciding if this should in fact be supported, or if we should throw a useful exception and ask the user to move the observer to the right device.

vkuzo commented on test/quantization/test_workflow_module.py:
```diff
@@ -482,6 +482,29 @@ def test_save_load_state_dict_script(self):
+        Ensure that the scale and zero_point returned by the observer
+        are on the same device as the input tensor.
+        """
+        observerList = [MinMaxObserver(),
+                        MovingAverageMinMaxObserver(),
+                        PerChannelMinMaxObserver(),
+                        MovingAveragePerChannelMinMaxObserver()]
```
hmm, looking at this test closely, I think this should fail unless the user moves the observers to CUDA explicitly.  I think the invariant to shoot for could be "observers should create tensors on the right device" and "observers should ensure that an input tensor has the same device as its internal buffers".    Thoughts?

supriyar committed:
Update on "[quant][qat] Ensure observer respects device affinity"


Summary:
Previosuly the scale and zero_point were returned on the CPU even if
the input tensor was on the GPU.
This is because `copy_()` doesn't respect the device when copying over the tensor.

Also fixed a bug where we were always setting the device to 'cuda' (irrespective of the device id)
in the calculate_qparams function

Test Plan:
python test/test_quantization.py TestObserver.test_observer_qparams_respects_device_affinity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24800495](https://our.internmc.facebook.com/intern/diff/D24800495)

[ghstack-poisoned]

vkuzo reviewed (approved):
lg, the test failure in `QuantizePerChannel4d` looks potentially related

supriyar committed:
Update on "[quant][qat] Ensure observer respects device affinity"


Summary:

fixed a bug where we were always setting the device to 'cuda' (irrespective of the device id)
in the calculate_qparams function

Fixes #46533

Test Plan:
python test/test_quantization.py TestObserver.test_observer_qparams_respects_device_affinity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24800495](https://our.internmc.facebook.com/intern/diff/D24800495)

[ghstack-poisoned]
